airflow/providers/sftp/hooks/sftp.py

```diff
@@ -166,16 +168,21 @@ def describe_directory(self, path: str) -> Dict[str, Dict[str, str]]:
             }
         return files
 
-    def list_directory(self, path: str) -> List[str]:
+    def list_directory(self, path: str, regex_pattern: Optional[str] = None) -> List[str]:
         """
         Returns a list of files on the remote system.
 
         :param path: full path to the remote directory to list
         :type path: str
+        :param regex_pattern: optional pattern to filter the remote_full_path files
+        :type: regex_pattern: Optional[str]
         """
         conn = self.get_conn()
-        files = conn.listdir(path)
-        return files
+        if regex_pattern:
+            pattern = re.compile(regex_pattern)
+            return [file for file in conn.listdir(path) if pattern.match(file)]
```

Comment:
would the usage of a [generator](https://docs.python.org/3/whatsnew/3.3.html#pep-380) a more efficient solution when encountering directories with many files?